File: test/language/sql_databases/cpp/MasterDatabaseTest.cpp
#include <cstdio>
#include <vector>
#include <string>

#include <sqlite3.h>

#include "gtest/gtest.h"

#include "sql_databases/MasterDatabase.h"

namespace sql_databases
{

class MasterDatabaseTest : public ::testing::Test
{
public:
    MasterDatabaseTest() : m_dbFileName("master_database_test.sqlite")
    {
        m_allTableNames.push_back("america");
        m_allTableNames.push_back("europe");
        m_allTableNames.push_back("usa");
        m_allTableNames.push_back("canada");
        m_allTableNames.push_back("slovakia");
        m_allTableNames.push_back("germany");
    }

    ~MasterDatabaseTest()
    {
        std::remove(m_dbFileName.c_str());
    }

protected:
    bool isTableInDb(zserio::SqlDatabase& database, const std::string& checkTableName)
    {
        sqlite3_stmt* statement;
        std::string sqlQuery = "SELECT name FROM sqlite_master WHERE type='table' AND name='" + checkTableName +
                "'";
        int result = sqlite3_prepare_v2(database.getConnection(), sqlQuery.c_str(), -1, &statement, NULL);
        if (result != SQLITE_OK)
            return false;

        result = sqlite3_step(statement);
        if (result == SQLITE_DONE || result != SQLITE_ROW)
        {
            sqlite3_finalize(statement);
            return false;
        }

        const unsigned char* readTableName = sqlite3_column_text(statement, 0);
        if (readTableName == NULL || checkTableName.compare(reinterpret_cast<const char*>(readTableName)) != 0)
        {
            sqlite3_finalize(statement);
            return false;
        }

        sqlite3_finalize(statement);

        return true;
    }

    const std::string m_dbFileName;

    typedef std::vector<std::string> StringVector;

    StringVector m_allTableNames;
};

TEST_F(MasterDatabaseTest, emptyConstructor)
{
    MasterDatabase database;
    database.open(m_dbFileName);
    ASSERT_TRUE(database.isOpen());
    database.close();
}

TEST_F(MasterDatabaseTest, externalConstructor)
{
    sqlite3 *externalConnection = NULL;
    const int result = sqlite3_open(m_dbFileName.c_str(), &externalConnection);
    ASSERT_EQ(SQLITE_OK, result);

    MasterDatabase database(externalConnection);
    database.createSchema();
    ASSERT_TRUE(database.isOpen());

    database.close();

    ASSERT_EQ(SQLITE_OK, sqlite3_close(externalConnection));
}

TEST_F(MasterDatabaseTest, fileNameConstructor)
{
    MasterDatabase database(m_dbFileName);
    ASSERT_TRUE(database.isOpen());
    database.close();
}

TEST_F(MasterDatabaseTest, externalOpen)
{
    sqlite3 *externalConnection = NULL;
    const int result = sqlite3_open(m_dbFileName.c_str(), &externalConnection);
    ASSERT_EQ(SQLITE_OK, result);

    MasterDatabase database;
    database.open(externalConnection);
    database.createSchema();
    ASSERT_TRUE(database.isOpen());

    database.close();

    ASSERT_EQ(SQLITE_OK, sqlite3_close(externalConnection));
}

TEST_F(MasterDatabaseTest, fileNameOpen)
{
    MasterDatabase database;
    database.open(m_dbFileName);
    ASSERT_TRUE(database.isOpen());
    database.close();
}

TEST_F(MasterDatabaseTest, createSchema)
{
    MasterDatabase database(m_dbFileName);

    for (StringVector::const_iterator it = m_allTableNames.begin(); it != m_allTableNames.end(); ++it)
        ASSERT_FALSE(isTableInDb(database, *it));
    database.createSchema();
    for (StringVector::const_iterator it = m_allTableNames.begin(); it != m_allTableNames.end(); ++it)
        ASSERT_TRUE(isTableInDb(database, *it));

    database.close();
}

TEST_F(MasterDatabaseTest, createSchemaWithoutRowIdBlackList)
{
    MasterDatabase database(m_dbFileName);

    for (StringVector::const_iterator it = m_allTableNames.begin(); it != m_allTableNames.end(); ++it)
        ASSERT_FALSE(isTableInDb(database, *it));
    const std::set<std::string> withoutRowIdTableNamesBlackList;
    database.createSchema(withoutRowIdTableNamesBlackList);
    for (StringVector::const_iterator it = m_allTableNames.begin(); it != m_allTableNames.end(); ++it)
        ASSERT_TRUE(isTableInDb(database, *it));

    database.close();
}

TEST_F(MasterDatabaseTest, deleteSchema)
{
    MasterDatabase database(m_dbFileName);
    database.createSchema();

    for (StringVector::const_iterator it = m_allTableNames.begin(); it != m_allTableNames.end(); ++it)
        ASSERT_TRUE(isTableInDb(database, *it));
    database.deleteSchema();
    for (StringVector::const_iterator it = m_allTableNames.begin(); it != m_allTableNames.end(); ++it)
        ASSERT_FALSE(isTableInDb(database, *it));

    database.close();
}

} // namespace sql_databases

File: test/language/sql_databases/java/sql_databases/MasterDatabaseTest.java
package sql_databases;

import static org.junit.Assert.*;

import java.io.File;
import java.net.URISyntaxException;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.HashSet;
import java.util.Properties;
import java.util.Set;

import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.Test;

import test_utils.FileUtil;
import test_utils.JdbcUtil;

import zserio.runtime.SqlDatabase;
import zserio.runtime.SqlDatabase.Mode;
import zserio.runtime.validation.ValidationReport;

public class MasterDatabaseTest
{
    @BeforeClass
    public static void init()
    {
        JdbcUtil.registerJdbc();
    }

    @Before
    public void setUp() throws SQLException
    {
        FileUtil.deleteFileIfExists(dbFile);
    }

    @Test
    public void fileNameConstructor() throws SQLException, URISyntaxException
    {
        database = new MasterDatabase(dbFile.toString());
        database.createSchema();
        checkDb(database);
        database.close();
    }

    @Test
    public void connectionConstructor() throws SQLException
    {
        final Properties connectionProps = new Properties();
        connectionProps.setProperty("flags", Mode.CREATE.toString());
        final String uriPath = "jdbc:sqlite:file:" + dbFile.toString();
        final Connection connection = DriverManager.getConnection(uriPath, connectionProps);

        database = new MasterDatabase(connection);
        database.createSchema();
        checkDb(database);
        connection.close();
    }

    @Test
    public void getDatabases() throws SQLException, URISyntaxException
    {
        database = new MasterDatabase(dbFile.toString());
        database.createSchema();

        final SqlDatabase[] databases = database.getDatabases();
        assertTrue(databases != null);
        assertEquals(NUM_ALL_DATABASES, databases.length);

        database.close();
    }

    @Test
    public void createSchema() throws SQLException, URISyntaxException
    {
        database = new MasterDatabase(dbFile.toString());

        for (String tableName : ALL_TABLE_NAMES)
            assertFalse(isTableInDb(tableName));
        database.createSchema();
        for (String tableName : ALL_TABLE_NAMES)
            assertTrue(isTableInDb(tableName));

        checkDb(database);
        database.close();
    }

    @Test
    public void createSchemaWithoutRowIdBlackList() throws SQLException, URISyntaxException
    {
        database = new MasterDatabase(dbFile.toString());

        for (String tableName : ALL_TABLE_NAMES)
            assertFalse(isTableInDb(tableName));
        final Set<String> withoutRowIdTableNamesBlackList = new HashSet<String>();
        database.createSchema(withoutRowIdTableNamesBlackList);
        for (String tableName : ALL_TABLE_NAMES)
            assertTrue(isTableInDb(tableName));

        checkDb(database);
        database.close();
    }

    @Test
    public void deleteSchema() throws SQLException, URISyntaxException
    {
        database = new MasterDatabase(dbFile.toString());
        database.createSchema();

        for (String tableName : ALL_TABLE_NAMES)
            assertTrue(isTableInDb(tableName));
        database.deleteSchema();
        for (String tableName : ALL_TABLE_NAMES)
            assertFalse(isTableInDb(tableName));

        database.close();
    }

    @Test
    public void validate() throws SQLException, URISyntaxException
    {
        database = new MasterDatabase(dbFile.toString());
        database.createSchema();
        checkDb(database);
        database.close();
    }

    private boolean isTableInDb(String checkTableName) throws SQLException
    {
        // check if database does contain table
        final String sqlQuery = "SELECT name FROM sqlite_master WHERE type='table' AND name='" +
                checkTableName + "'";

        final PreparedStatement statement = database.prepareStatement(sqlQuery);
        try
        {
            final ResultSet resultSet = statement.executeQuery();
            if (!resultSet.next())
                return false;

            // read table name
            final String readTableName = resultSet.getString(1);
            if (resultSet.wasNull() || !readTableName.equals(checkTableName))
                return false;
        }
        finally
        {
            statement.close();
        }

        return true;
    }

    private static void checkDb(MasterDatabase database) throws SQLException
    {
        final ValidationReport report = database.validate(null);
        assertEquals(NUM_ALL_TABLES, report.getNumberOfValidatedTables());
        assertEquals(0, report.getNumberOfValidatedRows());
        assertTrue(report.getTotalParameterProviderTime() <= report.getTotalValidationTime());
        assertTrue(report.getErrors().isEmpty());
    }

    private static final int NUM_ALL_DATABASES = 3;
    private static final int NUM_ALL_TABLES = 7;

    private static final String[] ALL_TABLE_NAMES = {"europe", "america", "usa", "canada", "slovakia",
        "czechia", "germany"};

    private static final String DB_FILE_NAME = "master_db_test.sqlite";

    private final File dbFile = new File(DB_FILE_NAME);
    private MasterDatabase database = null;
}
